# Experiment 6: One-Pass Stream Processing

## Aim

To read every value of a data stream exactly once, update the count, the running sum and the
running average immediately as each value arrives, and finally display the total elements
processed, the final sum, the final average and the execution time.

## Theory

One-pass stream processing means that every element of the stream is read only once and is
processed immediately on arrival. The complete data is never stored or scanned a second time, so
only a few summary variables such as count and total need to be kept in memory. From these two
variables the running average can be derived at any moment. This makes the method memory efficient
and suitable for continuous or very large data. The time module is used to record the start and end
time so that the execution time of the whole stream can be measured.

## Requirements

- Python 3.x
- Jupyter Notebook
- Time module

In [1]:
import time

stream_data = [5, 9, 11, 12, 15]

print("Stream data :", stream_data)

Stream data : [5, 9, 11, 12, 15]


In [2]:
count = 0
total = 0

start_time = time.time()

for value in stream_data:

    count += 1
    total += value
    running_average = total / count

    print("Received:", value)
    print("Running sum:", total)
    print("Running avg:", running_average)

end_time = time.time()

Received: 5
Running sum: 5
Running avg: 5.0
Received: 9
Running sum: 14
Running avg: 7.0
Received: 11
Running sum: 25
Running avg: 8.333333333333334
Received: 12
Running sum: 37
Running avg: 9.25
Received: 15
Running sum: 52
Running avg: 10.4


In [3]:
print("Total elements processed:", count)
print("Final sum:", total)
print("Final average:", total / count)
print("Execution time:", round(end_time - start_time, 6), "seconds")

Total elements processed: 5
Final sum: 52
Final average: 10.4
Execution time: 0.000663 seconds


## Interpretation

- The time module is needed to measure the execution time of the stream.
- Each value arrives sequentially and is processed immediately.
- Each element is processed exactly once.
- The running sum is updated immediately.
- The running average is calculated.
- It shows the total elements processed.
- It shows the final sum.
- It shows the final average.
- It shows the execution time.

## Result

The data stream was successfully processed in a single pass. Five elements were processed, giving
a final sum of 52 and a final average of 10.4, along with the execution time of the stream.